File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int solve(vector<int> & nums, int idx,vector<int>& dp){
        if(idx<0) return 0;
        if(dp[idx]!=-1) return dp[idx];
        return dp[idx] = max(solve(nums, idx-2, dp)+nums[idx], solve(nums, idx-1, dp));
    }
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), 0-1);
        return solve(nums,nums.size()-1, dp);
    }
};

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = [-1] * len(nums)
        def dfs(idx):
            if idx < 0: return 0
            if dp[idx] != -1: return dp[idx]
            dp[idx] = max(dfs(idx-2) + nums[idx], dfs(idx-1))  
            return dp[idx]
        return dfs(len(nums)-1)  